File: chapter13/code13_04.cpp
//始点sから終点tの経路が存在するかの判定問題をDFSで

#include <iostream>
#include <vector>

using namespace std;
using Graph = vector<vector<int>>;
vector<bool> seen;

void dfs(const Graph &G, int v)
{
    seen[v] = true;

    for (auto next_v : G[v])
    {
        if (seen[next_v])
            continue;
        dfs(G, next_v);
    }
}

int main()
{
    //頂点数 N　辺の数 M
    int N, M, s, t;
    cin >> N >> M >> s >> t;

    Graph G(N);
    for (int i = 0; i < M; ++i)
    {
        int a, b;
        cin >> a >> b;
        G[a].push_back(b);
        //有向グラフ
    }
    //seenを初期化
    seen.assign(N, false);
    dfs(G, s);

    if (seen[t])
        cout << "yes" << endl;
    else
        cout << "no" << endl;
}

File: chapter13/code13_04.py


seen = []


def dfs(G, s, seen):
    seen[s] = True

    for i in G:
        if i[0] == s:
            if seen[i[1]]:
                continue
            dfs(G, i[1], seen)


def main():
    N, M, s, t = map(int, input().split())

    Graph = [tuple(map(int, input().split())) for _ in range(M)]

    seen = [False]*N
    dfs(Graph, s, seen)

    if seen[t]:
        print("yes")
    else:
        print("no")


main()
